ramcache.py

```diff
@@ -0,0 +1,168 @@
+
+"""
+This module manages a distributed RAM cache as a global python dictionary in
+each AppEngine instance.  AppEngine can spin up new instances or kill old ones
+at any time.  Each instance's RAM cache is independent and might not have the
+same entries as found in the RAM caches of other instances.
+
+Each instance will do the work needed to compute a given RAM cache entry
+itself.  The values computed in a given instance will speed up future requests
+made to that instance only.
+
+When the user edits something in the app, the updated entity is stored in
+datastore.  Also, the singleton SharedInvalidate entity is updated with the
+timestamp of the change.  Every request handler must start processing a request
+by first calling SharedInvalidate.check_for_distributed_invalidation() which
+checks for any needed invalidations and clears RAM cache entries in
+that instance if needed.
+
+For now, there is only a single RAM cache per instance and when anything is
+invalidated, that entire RAM cache is completely cleared.  In the future,
+invalidations could be compartmentalized by RAM cache type, or even specific
+entity IDs.  Monorail uses that approach, but existing ChromeStatus code does
+not need it.
+
+Calling code must not mutate any value that is passed into set() or returned
+from get().  If calling code needs to mutate such objects, it should call
+copy.copy() or copy.deepcopy() to avoid unintentional cumulative mutations.
+
+Unlike memcache, this RAM cache has no concept of expiration time.  So,
+whenever a cached value would become invalid, it must be invalidated.
+"""
+
+import logging
+import time as time_module
+from google.appengine.ext import db
+
+
+global_cache = {}
+expires = {}
+
+# Whenever the cache would have more than this many items, some
+# random item is dropped, or the entire cache is cleared.
+# If our instances are killed by appengine for exceeding memory limits,
+# we can configure larger instances and/or reduce this value.
+MAX_CACHE_SIZE = 10000
+
+
+def set(key, value, time=None):
+  """Emulate the memcache.get() method using a RAM cache."""
+  if len(global_cache) + 1 > MAX_CACHE_SIZE:
+    popped_item = global_cache.popitem()
+    if popped_item[0] in expires:
+      del expires[popped_item[0]]
+  global_cache[key] = value
+  if time:
+    expires[key] = int(time_module.time()) + time
+
+
+def _check_expired(keys):
+  now = int(time_module.time())
+  for key in keys:
+    if key in expires and expires[key] < now:
+      del expires[key]
+      del global_cache[key]
+
+
+def get(key):
+  """Emulate the memcache.set() method using a RAM cache."""
+  _check_expired([key])
+  verb = 'hit' if key in global_cache else 'miss'
+  logging.info('cache %s for %r', verb, key)
+  return global_cache.get(key)
+
+
+def get_multi(keys):
+  """Emulate the memcache.get_multi() method using a RAM cache."""
+  _check_expired(keys)
+  return {
+      key: global_cache[key]
+      for key in keys
+      if key in global_cache}
```

Comment:
That closing bracket being on a line of its own would make this look less like something's got left out here IMO.